socket.js

```diff
@@ -212,66 +160,77 @@ function ReqSocket () {
   this.queues = [];
   this.callbacks = {};
 
+  this._deferredSends = [];
+
   if (this.ready && this.channel)
     this._setupConsumer();
   else
     this.once('ready', this._setupConsumer.bind(this));
 }
 
+ReqSocket.prototype._sendDeferredMessages = function () {
+  this.isConnected = true;
+
+  for (var i = 0; i < this._deferredSends.length; i++) {
+    var message = this._deferredSends[i];
+    this._sendNow(message.message, message.id);
+  }
+
+  this._deferredSends = [];
+};
+
 ReqSocket.prototype._setupConsumer = function () {
   var self = this;
   //
   // Remark: We create an ephemeral reply queue here to receive messages on
   //
-  this.channel.assertQueue('', {exclusive: true, autoDelete: true }, function (err, ok) {
-    if (err) { return self.emit('error', err); }
+  this.channel.assertQueue('', { exclusive: true, autoDelete: true }, function (err, ok) {
+    if (err) { return void self.emit('error', err); }
     self.replyTo = ok.queue;
-    self._canMaybeSend();
     self.channel.consume(ok.queue, self._consume.bind(self), { noAck: false, exclusive: true });
   });
 };
 
-ReqSocket.prototype._canMaybeSend = function () {
-  debug('canMaybeSend called');
-  if (!this.replyTo || !this.queues.length) return;
-  this._runDeferred('send');
-};
-
 ReqSocket.prototype._consume = function (msg) {
   if (msg === null) return debug('Req socket msg received is null');
   debug('Req socket received reply over ephemeral queue %s %j', this.replyTo, msg);
-  this.reply(msg);
+  this._handleReceipt(msg);
   this.channel.ack(msg);
 };
 
-ReqSocket.prototype.reply = function (msg) {
+ReqSocket.prototype._handleReceipt = function (msg) {
   var id = msg.properties.correlationId;
   var fn = this.callbacks[id];
+
+  // This means that we can fire and forget messages. Is this an intentional feature, or should
+  // this indicate that something went wrong?
   if (!fn) return debug('missing callback for %s', id);
 
   var message = this.parse(msg.content);
-  //
-  // Remark: since we can't really error should we just respond with the message?
-  //
+
   if (message.error && message.message) {
     var error = new Error(message.message);
-    return fn(error);
+    fn(error);
+    delete this.callbacks[id];
   }
   fn(null, message);
   delete this.callbacks[id];
 };
 
-ReqSocket.prototype.connect = function (destination, callback) {
+ReqSocket.prototype.connect = function (destination) {
+  if (!this.ready) {
+    this.deferredConnection = { queueName: destination };
+    return this;
+  }
   var self = this;
 
   this.channel.assertQueue(destination,
     { durable: this.options.persistent }, function (err, ok) {
-    if (err) return callback ? callback(err) : self.emit('error', err);
-    self.queues.push(ok.queue);
-    self._canMaybeSend();
-    self.emit('connect');
-    callback && callback();
-  });
+      if (err) return void self.emit('error', err);
+      self.queues.push(ok.queue);
+      self._sendDeferredMessages();
```

Comment:
We want to establish that our ephemeral queue has been established as well since they become part of the message we send to rabbit and enables the replies.